websocket_client.py

```diff
@@ -0,0 +1,463 @@
+
+    def __init__(self, sample_rate: int = 48000, num_channels: int = 1) -> None:
+        self._sample_rate = sample_rate
+        self._num_channels = num_channels
+        self._buffer: bytearray = bytearray()
+
+        def callback(outdata: np.ndarray, frame_count, time, status):
+            output_frame_size = len(outdata) * 2
+            next_frame = self._buffer[:output_frame_size]
+            self._buffer[:] = self._buffer[output_frame_size:]
+            if len(next_frame) < output_frame_size:
+                next_frame += b"\x00" * (output_frame_size - len(next_frame))
+            outdata[:] = np.frombuffer(next_frame, dtype="int16").reshape(
+                (frame_count, num_channels)
+            )
+
+        self._stream = sounddevice.OutputStream(
+            samplerate=sample_rate,
+            channels=num_channels,
+            callback=callback,
+            device=None,
+            dtype="int16",
+            blocksize=sample_rate // 100,
+        )
+        self._stream.start()
+        if not self._stream.active:
+            raise RuntimeError("Failed to start streaming output audio")
+
+    def write(self, chunk: bytes) -> None:
+        self._buffer.extend(chunk)
+
+    def drop_buffer(self) -> None:
+        self._buffer.clear()
+
+    async def close(self) -> None:
+        if self._stream:
+            self._stream.close()
+
+
+class LocalAudioSource:
+    """AudioSource that reads from the default microphone."""
+
+    def __init__(self, sample_rate=48000, channels=1):
+        self._sample_rate = sample_rate
+        self._num_channels = channels
+
+    async def stream(self) -> AsyncGenerator[bytes, None]:
+        queue: asyncio.Queue[bytes] = asyncio.Queue()
+        loop = asyncio.get_running_loop()
+
+        def callback(indata: np.ndarray, frame_count, time, status):
+            loop.call_soon_threadsafe(queue.put_nowait, indata.tobytes())
+
+        stream = sounddevice.InputStream(
+            samplerate=self._sample_rate,
+            channels=self._num_channels,
+            callback=callback,
+            device=None,
+            dtype="int16",
+            blocksize=self._sample_rate // 100,
+        )
+        with stream:
+            if not stream.active:
+                raise RuntimeError("Failed to start streaming input audio")
+            while True:
+                yield await queue.get()
+
+
+class WebsocketVoiceSession(pyee.asyncio.AsyncIOEventEmitter):
+    def __init__(self, join_url: str):
+        super().__init__()
+        self._state: Literal["idle", "listening", "thinking", "speaking"] = "idle"
+        self._pending_output = ""
+        self._url = join_url
+        self._socket = None
+        self._receive_task: asyncio.Task | None = None
+        self._send_audio_task = asyncio.create_task(
+            self._pump_audio(LocalAudioSource())
+        )
+        self._sink = LocalAudioSink()
+
+    async def start(self):
```

Comment:
"""
        Initialize and start the voice session.
        
        Establishes WebSocket connection and begins audio processing.
        """